File: sdks/python/src/scope_client/credentials.py
"""Credentials management for scope-client.

This module provides the credentials classes for SDK authentication.
The credential system is designed to be extensible for future auth schemes.

Example:
    >>> from scope_client import ClientCredentials
    >>>
    >>> # Create credentials explicitly
    >>> credentials = ClientCredentials(
    ...     org_id="my-org",
    ...     client_id="key_abc123",
    ...     client_secret="secret_xyz"
    ... )
    >>>
    >>> # Or load from environment variables
    >>> credentials = ClientCredentials.from_env()
"""

import os
import warnings
from typing import Any, Optional, Protocol, Union, runtime_checkable


@runtime_checkable
class CredentialsProtocol(Protocol):
    """Protocol defining the interface for all credential types.

    All credential implementations must provide these methods to be
    usable with the SDK.
    """

    @property
    def auth_type(self) -> str:
        """Return the authentication type identifier.

        Returns:
            String identifying the authentication type (e.g., 'client_credentials').
        """
        ...

    def validate(self) -> None:
        """Validate that all required credential fields are present.

        Raises:
            ConfigurationError: If required fields are missing.
        """
        ...

    def to_dict(self) -> dict[str, Any]:
        """Convert credentials to a dictionary representation.

        Sensitive fields should be redacted in the output.

        Returns:
            Dictionary representation of the credentials.
        """
        ...


class ClientCredentials:
    """Client credentials for SDK authentication.

    This is the primary authentication method for the Scope SDK,
    using organization ID, client ID, and client secret.

    Attributes:
        org_id: Organization identifier.
        client_id: Client ID for authentication.
        client_secret: Client secret for authentication.

    Note:
        The old parameter names ``api_key`` and ``api_secret`` are still accepted
        but deprecated. Use ``client_id`` and ``client_secret`` instead.

    Example:
        >>> credentials = ClientCredentials(
        ...     org_id="my-org",
        ...     client_id="key_abc123",
        ...     client_secret="secret_xyz"
        ... )
        >>> credentials.validate()  # Raises if invalid

        >>> # Load from environment
        >>> credentials = ClientCredentials.from_env()
    """

    org_id: Optional[str]
    client_id: Optional[str]
    client_secret: Optional[str]

    def __init__(
        self,
        org_id: Optional[str] = None,
        client_id: Optional[str] = None,
        client_secret: Optional[str] = None,
        *,
        api_key: Optional[str] = None,
        api_secret: Optional[str] = None,
    ) -> None:
        # Handle backward-compatible parameter names
        if api_key is not None:
            if client_id is not None:
                raise TypeError("Cannot specify both 'client_id' and 'api_key'")
            warnings.warn(
                "The 'api_key' parameter is deprecated. Use 'client_id' instead.",
                DeprecationWarning,
                stacklevel=2,
            )
            client_id = api_key

        if api_secret is not None:
            if client_secret is not None:
                raise TypeError("Cannot specify both 'client_secret' and 'api_secret'")
            warnings.warn(
                "The 'api_secret' parameter is deprecated. Use 'client_secret' instead.",
                DeprecationWarning,
                stacklevel=2,
            )
            client_secret = api_secret

        object.__setattr__(self, "org_id", org_id)
        object.__setattr__(self, "client_id", client_id)
        object.__setattr__(self, "client_secret", client_secret)

    def __setattr__(self, name: str, value: Any) -> None:
        raise AttributeError(
            f"cannot set '{name}' attribute on immutable {type(self).__name__} instance"
        )

    def __delattr__(self, name: str) -> None:
        raise AttributeError(
            f"cannot delete '{name}' attribute on immutable {type(self).__name__} instance"
        )

    @property
    def api_key(self) -> Optional[str]:
        """Backward-compatible accessor for client_id (deprecated)."""
        return self.client_id

    @property
    def api_secret(self) -> Optional[str]:
        """Backward-compatible accessor for client_secret (deprecated)."""
        return self.client_secret

    @property
    def auth_type(self) -> str:
        """Return the authentication type identifier.

        Returns:
            The string 'client_credentials'.
        """
        return "client_credentials"

    def validate(self) -> None:
        """Validate that all required credential fields are present.

        Raises:
            ConfigurationError: If org_id, client_id, or client_secret is missing.
        """
        from scope_client.errors import ConfigurationError

        if not self.org_id:
            raise ConfigurationError("org_id is required")
        if not self.client_id:
            raise ConfigurationError("client_id is required")
        if not self.client_secret:
            raise ConfigurationError("client_secret is required")

    def to_dict(self) -> dict[str, Any]:
        """Convert credentials to a dictionary with redacted secret.

        Returns:
            Dictionary with org_id, client_id, and redacted client_secret.
        """
        return {
            "auth_type": self.auth_type,
            "org_id": self.org_id,
            "client_id": self.client_id,
            "client_secret": "[REDACTED]" if self.client_secret else None,
        }

    @classmethod
    def from_env(cls) -> "ClientCredentials":
        """Create credentials from environment variables.

        Reads from:
            - SCOPE_ORG_ID: Organization identifier
            - SCOPE_CLIENT_ID: Client ID (falls back to SCOPE_API_KEY with deprecation warning)
            - SCOPE_CLIENT_SECRET: Client secret (falls back to SCOPE_API_SECRET with warning)

        Returns:
            ClientCredentials instance with values from environment.

        Example:
            >>> import os
            >>> os.environ["SCOPE_ORG_ID"] = "my-org"
            >>> os.environ["SCOPE_CLIENT_ID"] = "key_abc123"
            >>> os.environ["SCOPE_CLIENT_SECRET"] = "secret_xyz"
            >>> credentials = ClientCredentials.from_env()
            >>> credentials.org_id
            'my-org'
        """
        org_id = os.environ.get("SCOPE_ORG_ID")

        client_id = os.environ.get("SCOPE_CLIENT_ID")
        if client_id is None and os.environ.get("SCOPE_API_KEY") is not None:
            warnings.warn(
                "The 'SCOPE_API_KEY' environment variable is deprecated. "
                "Use 'SCOPE_CLIENT_ID' instead.",
                DeprecationWarning,
                stacklevel=2,
            )
            client_id = os.environ.get("SCOPE_API_KEY")

        client_secret = os.environ.get("SCOPE_CLIENT_SECRET")
        if client_secret is None and os.environ.get("SCOPE_API_SECRET") is not None:
            warnings.warn(
                "The 'SCOPE_API_SECRET' environment variable is deprecated. "
                "Use 'SCOPE_CLIENT_SECRET' instead.",
                DeprecationWarning,
                stacklevel=2,
            )
            client_secret = os.environ.get("SCOPE_API_SECRET")

        return cls(
            org_id=org_id,
            client_id=client_id,
            client_secret=client_secret,
        )

    def __repr__(self) -> str:
        """Return a string representation with redacted secret.

        Returns:
            String representation showing org_id and client_id but not secret.
        """
        return (
            f"ClientCredentials(org_id={self.org_id!r}, "
            f"client_id={self.client_id!r}, client_secret='[REDACTED]')"
        )

    def __eq__(self, other: object) -> bool:
        if not isinstance(other, ClientCredentials):
            return NotImplemented
        return (
            self.org_id == other.org_id
            and self.client_id == other.client_id
            and self.client_secret == other.client_secret
        )

    def __hash__(self) -> int:
        return hash((self.org_id, self.client_id, self.client_secret))


# Backward-compatible alias
ApiKeyCredentials = ClientCredentials

# Union type for all supported credential types
# This will be expanded as new auth schemes are added
Credentials = Union[ClientCredentials]

File: sdks/ruby/lib/scope_client/credentials.rb
# frozen_string_literal: true

module ScopeClient
  module Credentials
    # Base class for all credential types.
    # All credential implementations must inherit from this class
    # and implement the required methods.
    class Base
      # Returns the authentication type identifier.
      # @return [Symbol] The authentication type (e.g., :client_credentials)
      def auth_type
        raise NotImplementedError, "#{self.class} must implement #auth_type"
      end

      # Validates that all required credential fields are present.
      # @raise [ConfigurationError] If required fields are missing
      def validate!
        raise NotImplementedError, "#{self.class} must implement #validate!"
      end

      # Converts credentials to a hash representation.
      # Sensitive fields should be redacted in the output.
      # @return [Hash] Hash representation of the credentials
      def to_h
        raise NotImplementedError, "#{self.class} must implement #to_h"
      end
    end

    # Client credentials for SDK authentication.
    #
    # This is the primary authentication method for the Scope SDK,
    # using organization ID, client ID, and client secret.
    #
    # @example Create credentials explicitly
    #   credentials = ScopeClient::Credentials::ApiKey.new(
    #     org_id: 'my-org',
    #     client_id: 'key_abc123',
    #     client_secret: 'secret_xyz'
    #   )
    #
    # @example Create credentials from environment variables
    #   credentials = ScopeClient::Credentials::ApiKey.from_env
    #
    class ApiKey < Base
      attr_reader :org_id, :client_id, :client_secret

      # Creates a new ApiKey credentials instance.
      #
      # @param org_id [String, nil] Organization identifier
      # @param client_id [String, nil] Client ID for authentication
      # @param client_secret [String, nil] Client secret for authentication
      # @param api_key [String, nil] Deprecated: Use client_id instead
      # @param api_secret [String, nil] Deprecated: Use client_secret instead
      def initialize(org_id: nil, client_id: nil, client_secret: nil, api_key: nil, api_secret: nil)
        super()
        @org_id = org_id
        @client_id = resolve_credential('api_key', 'client_id', api_key, client_id)
        @client_secret = resolve_credential('api_secret', 'client_secret', api_secret, client_secret)
      end

      # Backward-compatible alias for client_id.
      # @deprecated Use {#client_id} instead
      # @return [String, nil] The client ID
      def api_key
        client_id
      end

      # Backward-compatible alias for client_secret.
      # @deprecated Use {#client_secret} instead
      # @return [String, nil] The client secret
      def api_secret
        client_secret
      end

      # Returns the authentication type identifier.
      # @return [Symbol] The symbol :client_credentials
      def auth_type
        :client_credentials
      end

      # Validates that all required credential fields are present.
      # @raise [ConfigurationError] If org_id, client_id, or client_secret is missing
      def validate!
        raise ConfigurationError, 'org_id is required' if @org_id.nil? || @org_id.to_s.empty?
        raise ConfigurationError, 'client_id is required' if @client_id.nil? || @client_id.to_s.empty?
        raise ConfigurationError, 'client_secret is required' if @client_secret.nil? || @client_secret.to_s.empty?
      end

      # Converts credentials to a hash with redacted secret.
      # @return [Hash] Hash with auth_type, org_id, client_id, and redacted client_secret
      def to_h
        {
          auth_type: auth_type,
          org_id: @org_id,
          client_id: @client_id,
          client_secret: @client_secret ? '[REDACTED]' : nil
        }
      end

      # Creates credentials from environment variables.
      #
      # Reads from:
      #   - SCOPE_ORG_ID: Organization identifier
      #   - SCOPE_CLIENT_ID: Client ID (falls back to SCOPE_API_KEY with deprecation warning)
      #   - SCOPE_CLIENT_SECRET: Client secret (falls back to SCOPE_API_SECRET with deprecation warning)
      #
      # @return [ApiKey] ApiKey instance with values from environment
      #
      # @example
      #   ENV['SCOPE_ORG_ID'] = 'my-org'
      #   ENV['SCOPE_CLIENT_ID'] = 'key_abc123'
      #   ENV['SCOPE_CLIENT_SECRET'] = 'secret_xyz'
      #   credentials = ScopeClient::Credentials::ApiKey.from_env
      #   credentials.org_id # => 'my-org'
      #
      def self.from_env
        new(
          org_id: ENV.fetch('SCOPE_ORG_ID', nil),
          client_id: resolve_env_var('SCOPE_CLIENT_ID', 'SCOPE_API_KEY'),
          client_secret: resolve_env_var('SCOPE_CLIENT_SECRET', 'SCOPE_API_SECRET')
        )
      end

      # Resolves an environment variable with fallback to a deprecated name.
      # @param new_name [String] The preferred env var name
      # @param old_name [String] The deprecated env var name
      # @return [String, nil] The resolved value
      def self.resolve_env_var(new_name, old_name)
        value = ENV.fetch(new_name, nil)
        return value if value

        value = ENV.fetch(old_name, nil)
        warn "[DEPRECATION] #{old_name} is deprecated. Use #{new_name} instead." if value
        value
      end

      private_class_method :resolve_env_var

      # Returns a string representation with redacted secret.
      # @return [String] String representation showing org_id and client_id but not secret
      def inspect
        "#<#{self.class.name} org_id=#{@org_id.inspect} client_id=#{@client_id.inspect} client_secret=[REDACTED]>"
      end

      private

      # Resolves a credential value, handling backward compatibility with deprecated names.
      # @param old_name [String] The deprecated parameter name
      # @param new_name [String] The preferred parameter name
      # @param old_value [Object, nil] The value passed via the deprecated parameter
      # @param new_value [Object, nil] The value passed via the preferred parameter
      # @return [Object, nil] The resolved value
      def resolve_credential(old_name, new_name, old_value, new_value)
        if old_value && new_value
          raise ArgumentError, "Cannot specify both #{old_name} and #{new_name}. Use #{new_name} instead."
        end

        if old_value
          warn "[DEPRECATION] #{old_name} is deprecated. Use #{new_name} instead."
          return old_value
        end

        new_value
      end
    end
  end
end
